common/utils/EventLogRiemannClient.js

```diff
@@ -104,6 +105,11 @@ class EventLogRiemannClient {
       if (appGuid !== undefined) {
         eventName = `${eventName}.app_guid.${appGuid}`;
       }
+
+      if (serviceId !== undefined) {
+        let serviceName = require('../models').catalog.getServiceName(serviceId);
```

Comment:
move require to the top of the file.

Reply:
Actually wanted to discuss this one with you , in case I make it a global require , then I get the following error in UT execution : 

```
C02V64NTHTD6:service-fabrik-broker i336605$ npm test
 
> service-fabrik-broker@3.1.1 test /Users/i336605/code/opensource/service-fabrik-broker
> ./test/init test
 
========================= BROKER ==========================
 
circuit breaker is disabled
/Users/i336605/code/opensource/service-fabrik-broker/data-access-layer/bosh/BoshDirectorClient.js:18
class BoshDirectorClient extends HttpClient {
                                 ^
 
TypeError: Class extends value undefined is not a constructor or null
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/data-access-layer/bosh/BoshDirectorClient.js:18:34)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/common/models/Plan.js:4:28)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/common/models/Service.js:4:14)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/common/models/catalog.js:6:17)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/common/models/index.js:3:19)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/common/utils/EventLogRiemannClient.js:9:17)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/common/utils/index.js:15:31)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/test/test_broker/mocks/azureClient.js:5:15)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/test/test_broker/mocks/index.js:6:21)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/test/test_broker/support/index.js:18:16)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at /Users/i336605/code/opensource/service-fabrik-broker/node_modules/mocha/bin/_mocha:311:3
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/i336605/code/opensource/service-fabrik-broker/node_modules/mocha/bin/_mocha:310:10)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Function.Module.runMain (module.js:665:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
npm ERR! Test failed.  See above for more details.
```

Reply:
This happens because we probably have a circular dependency. The best way to figure this out is by using a tool called `Madge` (https://www.npmjs.com/package/madge ).

Reply:
```
madge --circular broker
Processed 139 files (4.4s) (6 warnings)

× Found 11 circular dependencies!

1) ../common/utils/index.js > ../common/EventLogInterceptor.js
2) ../data-access-layer/eventmesh/index.js > ../data-access-layer/eventmesh/LockManager.js
3) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/BackupReaperJob.js
4) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/BnRStatusPollerJob.js
5) ../jobs/index.js > ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/BnRStatusPollerJob.js > ../managers/backup-manager/index.js > ../managers/backup-manager/BackupService.js
6) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/DbCollectionReaperJob.js
7) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/OperationStatusPollerJob.js
8) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ScheduleBackupJob.js
9) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ScheduledOobDeploymentBackupJob.js
10) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ServiceInstanceUpdateJob.js
11) lib/fabrik/index.js > lib/fabrik/Fabrik.js > lib/fabrik/DBManager.js > lib/fabrik/DirectorManager.js > lib/fabrik/DirectorInstance.js > ../jobs/index.js > ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ServiceInstanceUpdateJob.js

```

Reply:
```
madge --circular api-controllers
Processed 127 files (4.1s)

× Found 15 circular dependencies!

1) ../common/utils/index.js > ../common/EventLogInterceptor.js
2) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/BackupReaperJob.js
3) ../data-access-layer/eventmesh/index.js > ../data-access-layer/eventmesh/LockManager.js
4) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/BnRStatusPollerJob.js
5) ../jobs/index.js > ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/BnRStatusPollerJob.js > ../managers/backup-manager/index.js > ../managers/backup-manager/BackupService.js
6) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/DbCollectionReaperJob.js
7) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/OperationStatusPollerJob.js
8) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ScheduleBackupJob.js
9) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ScheduledOobDeploymentBackupJob.js
10) ../broker/lib/fabrik/DBManager.js > ../broker/lib/fabrik/DirectorManager.js > ../broker/lib/fabrik/DirectorInstance.js > ../jobs/index.js > ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ServiceInstanceUpdateJob.js > ../broker/lib/fabrik/index.js > ../broker/lib/fabrik/Fabrik.js
11) ../broker/lib/fabrik/DirectorManager.js > ../broker/lib/fabrik/DirectorInstance.js > ../jobs/index.js > ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ServiceInstanceUpdateJob.js > ../broker/lib/fabrik/index.js > ../broker/lib/fabrik/Fabrik.js
12) ../broker/lib/fabrik/DirectorManager.js > ../broker/lib/fabrik/DirectorInstance.js > ../jobs/index.js > ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ServiceInstanceUpdateJob.js > ../broker/lib/fabrik/index.js > ../broker/lib/fabrik/Fabrik.js > ../broker/lib/fabrik/DirectorTaskPoller.js
13) ../broker/lib/fabrik/DirectorManager.js > ../broker/lib/fabrik/DirectorInstance.js > ../jobs/index.js > ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ServiceInstanceUpdateJob.js > ../broker/lib/fabrik/index.js > ../broker/lib/fabrik/Fabrik.js > ../broker/lib/fabrik/FabrikStatusPoller.js
14) ../broker/lib/fabrik/DirectorManager.js > ../broker/lib/fabrik/DirectorInstance.js > ../jobs/index.js > ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ServiceInstanceUpdateJob.js > ../broker/lib/fabrik/index.js > ../broker/lib/fabrik/Fabrik.js > ../broker/lib/fabrik/FabrikStatusPoller.js > ../broker/lib/fabrik/ServiceFabrikOperation.js
15) ../jobs/ScheduleManager.js > ../jobs/JobFabrik.js > ../jobs/ServiceInstanceUpdateJob.js
```

Reply:
@abh1kg : thanks for this info !

@subhankarc / @amitmalav / @muralisc : Is this circular dependency required ? I see its introduced with sf2.0 restructuring - https://github.com/cloudfoundry-incubator/service-fabrik-broker/pull/330

Reply:
Just checked with 6b branch, the circular dependencies were there before as well. I think the paths have changed since restructuring. 

```
$ madge --circular .
Processed 282 files (7.9s) (1 warning)

✖ Found 13 circular dependencies!

1) common/utils/index.js > common/EventLogInterceptor.js
2) broker/lib/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/BackupReaperJob.js
3) broker/lib/fabrik/DirectorManager.js > broker/lib/fabrik/DirectorInstance.js > broker/lib/jobs/index.js > broker/lib
/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/BnRStatusPollerJob.js
4) broker/lib/fabrik/DirectorManager.js > broker/lib/fabrik/DirectorInstance.js > broker/lib/jobs/index.js > broker/lib
/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/BnRStatusPollerJob.js > broker/lib/fabrik/Ser
viceFabrikOperation.js
5) broker/lib/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/BnRStatusPollerJob.js
6) eventmesh/index.js > eventmesh/LockManager.js
7) broker/lib/jobs/index.js > broker/lib/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/BnRSt
atusPollerJob.js > managers/backup-manager/index.js > managers/backup-manager/BackupService.js
8) broker/lib/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/DbCollectionReaperJob.js
9) broker/lib/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/OperationStatusPollerJob.js
10) broker/lib/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/ScheduleBackupJob.js
11) broker/lib/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/ScheduledOobDeploymentBackupJob
.js
12) broker/lib/fabrik/index.js > broker/lib/fabrik/Fabrik.js > broker/lib/fabrik/DBManager.js > broker/lib/fabrik/Direc
torManager.js > broker/lib/fabrik/DirectorInstance.js > broker/lib/jobs/index.js > broker/lib/jobs/ScheduleManager.js >
 broker/lib/jobs/JobFabrik.js > broker/lib/jobs/ServiceInstanceUpdateJob.js
13) broker/lib/jobs/ScheduleManager.js > broker/lib/jobs/JobFabrik.js > broker/lib/jobs/ServiceInstanceUpdateJob.js
```

@kamath-prasad May be we can take this up after the introduction of the managers. Is that fine?

Reply:
@subhankarc : Oh ok :) Sorry for my ignorance ! Just saw 1 commit on Plan.js :) ..

Yes, it should be fine.

@jagadish-kb : is it fine that we have the lazy initialisation for now ?

Reply:
I don't see this file in the list. Do you see it?

Reply:
Even I dont see it :) Lets discuss in person and see if it's really due to circular dependency or not !

Reply:
Oh...It is because plan.js refers to BoshDirectorClient which again refers to utils for HTTPClient.